lib/src/main/java/io/ably/lib/rest/Auth.java

```diff
@@ -128,16 +128,53 @@ public AuthOptions(String key) throws AblyException {
 		}
 
 		/**
-		 * Internal
+		 * Internal, when AuthOptions are provided, the values of each attribute are not merged,
+		 * but instead replace all corresponding values, even when null.
+		 *
+		 * should be used for:
+		 * - {@link Auth#requestToken(TokenParams, AuthOptions)} (Spec:RSA8e),
+		 * - {@link Auth#createTokenRequest(AuthOptions, TokenParams)} (Spec:RSA9h),
+		 * - {@link Auth#authorise(AuthOptions, TokenParams)} (Spec:RSA10j).
+		 *
+		 * <p>
+		 * Spec: RSA8e, RSA9h, RSA10j
+		 * </p>
+		 *
+		 * @param options if the object is null, the client library configured defaults are used.
+		 * @return superseded object
 		 */
-		public AuthOptions merge(AuthOptions defaults) {
-			if(authCallback == null) authCallback = defaults.authCallback;
-			if(authUrl == null) authUrl = defaults.authUrl;
-			if(key == null) key = defaults.key;
-			if(authHeaders == null) authHeaders = defaults.authHeaders;
-			if(authParams == null) authParams = defaults.authParams;
-			queryTime = queryTime & defaults.queryTime;
-			return this;
+		public AuthOptions replace(AuthOptions options) {
+			AuthOptions copy;
+			if (this instanceof ClientOptions) {
```

Comment:
The implementation choice to extend ClientOptions with AuthOptions is fine, perhaps composition would have been better (we use that primarily with Ruby), but I don't think that in itself is a problem.

The problem is that you're doing checks for a ClientOptions argument which we do not support in our API.  The API docs and IDL is all that we care about, and you have added code to accept ClientOptions and work with that specifically when that is not supported.

So if ClientOptions adheres to the AuthOptions spec, and that is passed in to the `authorise` method, then that is up to the developer even though it would be weird.  But we should then only be treating it as a AuthOptions object and by specifying the type as AuthOptions for this argument, we don't care and should not care if it's FooOptions but inherits from AuthOptions as we will only use the interface defined by AuthOptions.

Following the current logic in this PR, we'd need to support potentially any other objects that may inherit from AuthOptions, but that is never necessary and entire point of having interfaces.

So please stick to the spec, the types defined, and only handle those types based on the interfaces specified in the types.  That way the compiler will ensure the types you work with match the requirements of your code.